core/server/common/src/main/java/alluxio/master/journal/raft/BufferedJournalApplier.java

```diff
@@ -0,0 +1,298 @@
+  private static final Logger LOG = LoggerFactory.getLogger(BufferedJournalApplier.class);
+
+  /** Resume will acquire exclusive lock when buffer has less than this value. */
+  private static final int RESUME_LOCK_BUFFER_SIZE_WATERMARK = 100;
+
+  /** Resume will acquire exclusive lock when can't finish after this much time. */
+  private static final int RESUME_LOCK_TIME_LIMIT_MS = 30000;
+
+  /** Journals managed by this applier. */
+  private final Map<String, RaftJournal> mJournals;
+  /** A supplier of journal sinks for this applier. */
+  private final Supplier<Set<JournalSink>> mJournalSinks;
+
+  /** The last sequence applied. */
+  private long mLastAppliedSequence = -1;
+
+  /** Whether this state machine is suspended. */
+  @GuardedBy("mStateLock")
+  private boolean mSuspended = false;
+
+  /** Whether a resume() call is in progress. */
+  @GuardedBy("mStateLock")
+  private boolean mResumeInProgress = false;
+
+  /** Buffer to use during suspend period. */
+  private Queue<Journal.JournalEntry> mSuspendBuffer = new ConcurrentLinkedQueue();
+
+  /** Used to store latest catch-up thread. */
+  private AbstractCatchupThread mCatchupThread;
+
+  /** Used to synchronize buffer state. */
+  private ReentrantLock mStateLock = new ReentrantLock(true);
+
+  /**
+   * Creates a buffered applier over given journals.
+   *
+   * @param journals journals
+   * @param journalSinks journal sinks
+   */
+  public BufferedJournalApplier(Map<String, RaftJournal> journals,
+      Supplier<Set<JournalSink>> journalSinks) {
+    mJournals = journals;
+    mJournalSinks = journalSinks;
+  }
+
+  /**
+   * Processes given journal entry for applying. Entry could be applied or buffered based on
+   * buffer's state.
+   *
+   * @param journalEntry the journal entry
+   */
+  public void processJournalEntry(Journal.JournalEntry journalEntry) {
+    try (LockResource stateLock = new LockResource(mStateLock)) {
+      if (mSuspended) {
+        // New entry submissions are monitored by catch-up threads.
+        synchronized (mSuspendBuffer) {
+          mSuspendBuffer.offer(journalEntry);
+          mSuspendBuffer.notifyAll();
+        }
+      } else {
+        applyToMaster(journalEntry);
+      }
+    }
+  }
+
+  /**
+   * @return {@code true} if this applier was suspended
+   */
+  public boolean isSuspended() {
+    try (LockResource stateLock = new LockResource(mStateLock)) {
+      return mSuspended;
+    }
+  }
+
+  /**
+   * Suspend the applier.
+   *
+   * After this call, journal entries will be buffered until {@link #resume()} or
+   * {@link #catchup(long)} is called.
+   *
+   * @throws IOException
+   */
+  public void suspend() throws IOException {
+    try (LockResource stateLock = new LockResource(mStateLock)) {
+      Preconditions.checkState(!mSuspended, "Already suspended");
+      LOG.info("Suspending state machine.");
+      mSuspended = true;
+      LOG.info("Suspended state machine at sequence: {}", mLastAppliedSequence);
```

Comment:
The previous log message can be removed since there is trivial logic in between.